modules/assim.sequential/R/downscale_function.R

```diff
@@ -1,88 +1,237 @@
+  
+  # Convert site coordinates to SpatVector
+  site_coordinates <- terra::vect(preprocessed$site_coordinates, geom = c("lon", "lat"), crs = "EPSG:4326")
+  
+  # Extract predictors from covariates raster using site coordinates
+  predictors <- base::as.data.frame(terra::extract(covariates, site_coordinates, ID = FALSE))
+  
+  # Dynamically get covariate names
+  covariate_names <- base::names(predictors)
+  
+  # Create a single data frame with all predictors and ensemble data
+  full_data <- base::cbind(carbon_data, predictors)
+  
+  # Split the observations into training and testing sets
+  if (!base::is.null(seed)) {
+    base::set.seed(seed)  # Only set seed if provided
   }
+  sample <- base::sample(1:base::nrow(full_data), size = base::round(0.75 * base::nrow(full_data)))
+  train_data <- full_data[sample, ]
+  test_data <- full_data[-sample, ]
+  
+  # Prepare data for both RF and CNN
+  x_data <- base::as.matrix(full_data[, covariate_names])
+  y_data <- base::as.matrix(carbon_data)
   
-  # Generate predictions (maps) for each ensemble member using the trained models
-  maps <- list(ncol(rf_output))
-  for (i in 1:length(rf_output)) {
-    maps[[i]] <- terra::predict(object = covariates,
-                                model = rf_output[[i]],na.rm = T)
+  # Calculate scaling parameters from all data
+  scaling_params <- base::list(
+    mean = base::colMeans(x_data),
+    sd = base::apply(x_data, 2, stats::sd)
+  )
+  
+  # Normalize the data
+  x_data_scaled <- base::scale(x_data, center = scaling_params$mean, scale = scaling_params$sd)
+  
+  # Split into training and testing sets
+  x_train <- x_data_scaled[sample, ]
+  x_test <- x_data_scaled[-sample, ]
+  y_train <- y_data[sample, ]
+  y_test <- y_data[-sample, ]
+  
+  # Initialize lists for outputs
+  models <- base::list()
+  maps <- base::list()
+  predictions <- base::list()
+  
+  if (model_type == "rf") {
+    for (i in base::seq_along(carbon_data)) {
+      ensemble_col <- base::paste0("ensemble", i)
+      formula <- stats::as.formula(base::paste(ensemble_col, "~", base::paste(covariate_names, collapse = " + ")))
+      models[[i]] <- randomForest::randomForest(formula,
+                                                data = train_data,
+                                                ntree = 1000,
+                                                na.action = stats::na.omit,
+                                                keep.forest = TRUE,
+                                                importance = TRUE)
+      
+      maps[[i]] <- terra::predict(covariates, model = models[[i]], na.rm = TRUE)
+      predictions[[i]] <- stats::predict(models[[i]], test_data)
+    }
+  } else if (model_type == "cnn") {
+    x_train <- array_reshape(x_train, c(base::nrow(x_train), 1, base::ncol(x_train)))
+    x_test <- array_reshape(x_test, c(base::nrow(x_test), 1, base::ncol(x_test)))
+    
+    for (i in base::seq_along(carbon_data)) {
+      model <- keras_model_sequential() |>
```

Comment:
still missing namespaces on array_reshape, keras functions, etc. associated with CNN model

Reply:
addressed in [38c9e7a](https://github.com/PecanProject/pecan/pull/3321/commits/38c9e7a31e6417bdbce5cdf421647a239982a86f) commit